### ISSUE TITLE
Memory Profile not deserialized properly in PowerShell workflows, encountering wrong deserialized object type

### ISSUE THREAD

**Issue by slapointe** [Azure/azure-powershell#392]
Windows Server 2012 R2
PowerShell cmdlets 0.8.16 .. 0.9.1

I saw that 0.8.16 made AzureProfile serializable for workflows. It is not working for me though.

Inside a Workflow I get an error when passing the type Microsoft.Azure.Common.Authentication.Models.AzureProfile.

---------------------------

I get the following error:

Cannot convert the ... of type "Deserialized.Microsoft.Azure.Common.Authentication.Models.AzureProfile" to type  "Microsoft.Azure.Common.Authentication.Models.AzureProfile"."

---------------------------

Here is a sample that reproduce the problem:

$subscriptionId = $null # Enter subscriptionId here
$cred= get-credential
$azureProfile = New-AzureProfile -SubscriptionId $subscriptionId -Credential $cred

workflow Test-IssueWithWorkflowAndAzureProfile {
param(
    $Profile
)
$VerbosePreference = 'Continue'

Write-Verbose 'From workflow'
 $Profile
 Select-AzureProfile -Profile $Profile

 InlineScript {
        Import-Module Azure
        $VerbosePreference = 'Continue'
        Write-Verbose 'From inline'
        Select-AzureProfile -Profile $using:Profile
        Get-AzureVM
 }
}

Test-IssueWithWorkflowAndAzureProfile -Profile $azureProfile

**Comment by jodoglevy** [comment]
Hey @devigned - this isn't an issue with the Azure Automation cmdlets (although it will affect all Azure cmdlets' ability to be used with stateless profiles in Azure Automation) - may want to reassign.

**Comment by devigned** [comment]
Thank you, @jodoglevy.

@matt-gibbs, thoughts?

**Comment by stankovski** [comment]
@slapointe can you help us repro this issue without workflow? We already have tests that do serialization and deserialization of AzureProfile that are passing: https://github.com/Azure/azure-sdk-for-net/blob/master/src/Authentication/Common.Authentication.Tests/ProfileTests.cs#L80

**Comment by slapointe** [comment]
Right now I don't have a way to reproduce this without a PS workflow. I can take some time next week to see If I can find something.

**Comment by slapointe** [comment]
I was suspecting for a while now that PowerShell workflows were the problem here and I found this while searching today

**SMA / Azure Automation gotcha # 2**

I have a script that runs in PowerShell, but not in Workflow (SMA or Azure Automation).  My script uses piping to pass objects from one cmdlet to another.  Why do I get an error when using PowerShell Workflow? The cmdlets are failing with an error message similar to the following (this was generated when trying to use an Azure VM object as a parameter):

Cannot bind parameter 'VM'. Cannot convert the "Microsoft.WindowsAzure.Commands.ServiceManagement.Model.PersistentVM" value of type "Deserialized.Microsoft.WindowsAzure.Commands.ServiceManagement.Model.PersistentVM" to type "Microsoft.WindowsAzure.Commands.ServiceManagement.Model.IPersistentVM". (Cannot bind parameter 'VM'. Cannot convert the "Microsoft.WindowsAzure.Commands.ServiceManagement.Model.PersistentVM" value of type "Deserialized.Microsoft.WindowsAzure.Commands.ServiceManagement.Model.PersistentVM" to type "Microsoft.WindowsAzure.Commands.ServiceManagement.Model.IPersistentVM". (Cannot convert the "Microsoft.WindowsAzure.Commands.ServiceManagement.Model.PersistentVM" value of type "Deserialized.Microsoft.WindowsAzure.Commands.ServiceManagement.Model.PersistentVM" to type "Microsoft.WindowsAzure.Commands.ServiceManagement.Model.IPersistentVM".))
 
This is because PS Workflow stores complex objects in a deserialized format in order to enable checkpointing (workflow must be able to persist the data to be able to resume).  You are passing a complex object as a parameter to a PowerShell cmdlet and it received [Deserialized.ParameterType], when it was expecting [ParameterType].

Wrap any cmdlets that generate complex objects that are being passed as parameters to another cmdlet in InlineScript (both cmdlets should be wrapped in same inlinescript) to solve this problem.  

Taken from: http://social.technet.microsoft.com/wiki/contents/articles/25123.powershell-workflow-gotchas.aspx

In order to fix this, I can do right now:
```powershell
$subscriptionId = $null # Enter subscriptionId here
$cred= get-credential

workflow Test-IssueWithWorkflowAndAzureProfile 
{
    param(
        [string]
        $SubscriptionId,
        [PSCredential]
        $Credential
    )
    $VerbosePreference = 'Continue'

    InlineScript {
        Import-Module -Name Azure
        $VerbosePreference = 'Continue'
        Write-Verbose -Message 'From inline'
        $azureProfile = New-AzureProfile -SubscriptionId $using:SubscriptionId -Credential $using:Credential
        Select-AzureProfile -Profile $azureProfile

        $azureProfile = New-AzureProfile -SubscriptionId $using:SubscriptionId -Credential $using:Credential
        Select-AzureProfile -Profile $azureProfile

        Get-AzureVM
    }
}

Test-IssueWithWorkflowAndAzureProfile -SubscriptionId $subscriptionId -Credential $cred
```

It will works that way (note the double authentication, still an issue with memory profiles, see #362). This is horrible in worflows because a single call to New-AzureProfile for my credential in Azure take over 30 seconds to complete because of the number of subscription I have. This requires to re-authenticate in each *InlineScript*. I cannot even think about the delay in the scripts that this introduce. We need to think of something better. 

**Suggestion**
We could have a new cmdlet named Save-AzureProfile that serialize the AzureProfile including the auth token in a less complex object that we could pass to a new parameterset of Select-AzureProfile.

```powershell
$subscriptionId = $null # Enter subscriptionId here
$cred= get-credential
$azureProfile = New-AzureProfile -SubscriptionId $subscriptionId -Credential $cred
$savedProfile = Save-AzureProfile -Profile $azureProfile #result can be a scrambled string or binary representation in a simple object like a string

workflow Test-IssueWithWorkflowAndAzureProfile 
{
    param(
        $savedAzureProfile
    )
    $VerbosePreference = 'Continue'

    InlineScript {
        Import-Module -Name Azure
        $VerbosePreference = 'Continue'
        Write-Verbose -Message 'From inline'

        Select-AzureProfile -Profile $using:savedAzureProfile

        Get-AzureVM
    }
}

Test-IssueWithWorkflowAndAzureProfile -Profile $savedProfile
```

More elegant and only one authentication, for one or many InlineScript.

**Comment by jodoglevy** [comment]
@slapointe I don't think your suggestion would be needed if AzureProfile correctly deserialized, right? I think that's the proper solution.

I agree on just using a PowerShell Workflow to test. @stankovski why can't a PowerShell Workflow be used as a test case?

AzureProfile working correctly in PowerShell Workflow is an important scenario for Azure Automation, it'd be great to get this fixed.

**Comment by slapointe** [comment]
@jodoglevy I completely agree, we don't need this if we find the right way to get the object with the right type after deserialization.

I looked around a bit using reflector which lead me to PowerShell type conversion which I didn't know about. I found this article that explain the "magic" of PowerShell type conversion:

http://blogs.msdn.com/b/powershell/archive/2013/06/11/understanding-powershell-s-type-conversion-magic.aspx

I'll do some tests with **10 - TypeConverter conversion** because I think it is easier to test, but I'm pretty sure the're is a way to use a conversion strategy above that (the lower numbers), which would feel more natural and won't requires extra registration.
